### Pandas

Useful for data processing and analysis


#### Pandas Dataframe

Pandas Dataframe is two-dimensional tabular data structure with labelled axes(rows and columns

In [6]:
import pandas as pd #importing pandas library

In [7]:
from sklearn.datasets import fetch_california_housing

In [8]:
# Load the California housing dataset
california_housing = fetch_california_housing()

In [9]:
print(type(california_housing))

<class 'sklearn.utils._bunch.Bunch'>


In [10]:
print(california_housing)

{'data': array([[   8.3252    ,   41.        ,    6.98412698, ...,    2.55555556,
          37.88      , -122.23      ],
       [   8.3014    ,   21.        ,    6.23813708, ...,    2.10984183,
          37.86      , -122.22      ],
       [   7.2574    ,   52.        ,    8.28813559, ...,    2.80225989,
          37.85      , -122.24      ],
       ...,
       [   1.7       ,   17.        ,    5.20554273, ...,    2.3256351 ,
          39.43      , -121.22      ],
       [   1.8672    ,   18.        ,    5.32951289, ...,    2.12320917,
          39.43      , -121.32      ],
       [   2.3886    ,   16.        ,    5.25471698, ...,    2.61698113,
          39.37      , -121.24      ]]), 'target': array([4.526, 3.585, 3.521, ..., 0.923, 0.847, 0.894]), 'frame': None, 'target_names': ['MedHouseVal'], 'feature_names': ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude'], 'DESCR': '.. _california_housing_dataset:\n\nCalifornia Housing dataset\n-

In [11]:
#pandas dataframe
boston_df = pd.DataFrame(california_housing,columns=california_housing.feature_names)

In [12]:
boston_df.head()

Empty DataFrame
Columns: [MedInc, HouseAge, AveRooms, AveBedrms, Population, AveOccup, Latitude, Longitude]
Index: []

In [14]:
boston_df.shape

(0, 8)

In [15]:
#importing the data from csv file to a pandas dataframe
#creating a pandas dataframe
df = pd.read_csv(r"C:\Users\bhara\Downloads\boston.csv")

In [16]:
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

##### loading the data from excel file to pandas dataframe
pd.read_excel('filepath')

##### Exporting a dataframe to a csv file

boston_df.to_csv('boston.csv') # here boston refers to file name

##### Exporting pandas dataframme to excel file

filename.to_excel('filename')

#### Creating a dataframe with random values


In [18]:
import numpy as np

In [19]:
random_df = pd.DataFrame(np.random.rand(20,10)) # this will create 20 rows and 10 columns of random df

In [20]:
random_df.head()

0         1         2         3         4         5         6  \
0  0.029510  0.271285  0.675159  0.818624  0.012923  0.093850  0.998955   
1  0.523259  0.093292  0.161122  0.649982  0.782256  0.731629  0.403015   
2  0.998978  0.830159  0.707968  0.793325  0.594726  0.383558  0.939518   
3  0.287277  0.368109  0.117611  0.332439  0.183924  0.837030  0.759880   
4  0.386477  0.121961  0.775433  0.739157  0.725661  0.189625  0.180401   

          7         8         9  
0  0.285294  0.666217  0.480099  
1  0.746489  0.738604  0.877838  
2  0.810470  0.429319  0.154489  
3  0.623414  0.595169  0.812396  
4  0.796641  0.276394  0.069458

In [21]:
random_df.shape

(20, 10)

##### Inspecting a dataframe

In [22]:
#finding the number of rows and columns
df = pd.read_csv(r"C:\Users\bhara\Downloads\boston.csv")

In [23]:
df.shape #finding the rows and columns 

(506, 14)

In [24]:
#first 5 rows of the dataset
df.head(5)

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

In [25]:
#last 5 rows of the dataset
df.tail(5)

CRIM   ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
501  0.06263  0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273.0   
502  0.04527  0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273.0   
503  0.06076  0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273.0   
504  0.10959  0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273.0   
505  0.04741  0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273.0   

     PTRATIO       B  LSTAT  MEDV  
501     21.0  391.99   9.67  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9

In [26]:
# information about the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    int64  
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB


#### total we have 506 value and 13 columns
#### non-null values means no missing values
#### dtypes = datatypes
#### file size it give

In [27]:
# finding the number of missing value
df.isnull().sum()

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64

#### Here there are no missing values

In [ ]:
C:\Users\bhara\Downloads

In [31]:
db_df = pd.read_csv(r"C:\Users\bhara\Downloads\diabetes.csv")

In [32]:
db_df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [33]:
#counting the values based on the labels
db_df.value_counts('Outcome')  

Outcome
0    500
1    268
dtype: int64

#### specific values on the columns

In [35]:
#grouping the data based on the mean
db_df.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

#### here we have two outcome

for all the '0' we have mean value
for all the '1 ' we have mean value

### Statistical measures

Here we will use the boston dataframe

In [36]:
df.count()

CRIM       506
ZN         506
INDUS      506
CHAS       506
NOX        506
RM         506
AGE        506
DIS        506
RAD        506
TAX        506
PTRATIO    506
B          506
LSTAT      506
MEDV       506
dtype: int64

In [37]:
#mean value column wise
df.mean()

CRIM         3.613524
ZN          11.363636
INDUS       11.136779
CHAS         0.069170
NOX          0.554695
RM           6.284634
AGE         68.574901
DIS          3.795043
RAD          9.549407
TAX        408.237154
PTRATIO     18.455534
B          356.674032
LSTAT       12.653063
MEDV        22.532806
dtype: float64

In [38]:
# standard deviation column wise
df.std()

CRIM         8.601545
ZN          23.322453
INDUS        6.860353
CHAS         0.253994
NOX          0.115878
RM           0.702617
AGE         28.148861
DIS          2.105710
RAD          8.707259
TAX        168.537116
PTRATIO      2.164946
B           91.294864
LSTAT        7.141062
MEDV         9.197104
dtype: float64

In [39]:
# minimum value
df.min()

CRIM         0.00632
ZN           0.00000
INDUS        0.46000
CHAS         0.00000
NOX          0.38500
RM           3.56100
AGE          2.90000
DIS          1.12960
RAD          1.00000
TAX        187.00000
PTRATIO     12.60000
B            0.32000
LSTAT        1.73000
MEDV         5.00000
dtype: float64

In [40]:
df.max()

CRIM        88.9762
ZN         100.0000
INDUS       27.7400
CHAS         1.0000
NOX          0.8710
RM           8.7800
AGE        100.0000
DIS         12.1265
RAD         24.0000
TAX        711.0000
PTRATIO     22.0000
B          396.9000
LSTAT       37.9700
MEDV        50.0000
dtype: float64

In [41]:
# we can get all the statistical measures in dataframe
df.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT        MEDV  
count  506.000000  506.000000  
mean    12.653063   22.532806  
std      7.141062    9.197104  
min      1.730000    5.000000  
25%      6.950000   17.025000  
50%     11.360000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

#### Manipulating a dataframe

In [47]:
# adding a column to a dataframe
from sklearn.datasets import fetch_california_housing

In [48]:
df_boston = fetch_california_housing()

In [50]:
df_boston

{'data': array([[   8.3252    ,   41.        ,    6.98412698, ...,    2.55555556,
           37.88      , -122.23      ],
        [   8.3014    ,   21.        ,    6.23813708, ...,    2.10984183,
           37.86      , -122.22      ],
        [   7.2574    ,   52.        ,    8.28813559, ...,    2.80225989,
           37.85      , -122.24      ],
        ...,
        [   1.7       ,   17.        ,    5.20554273, ...,    2.3256351 ,
           39.43      , -121.22      ],
        [   1.8672    ,   18.        ,    5.32951289, ...,    2.12320917,
           39.43      , -121.32      ],
        [   2.3886    ,   16.        ,    5.25471698, ...,    2.61698113,
           39.37      , -121.24      ]]),
 'target': array([4.526, 3.585, 3.521, ..., 0.923, 0.847, 0.894]),
 'frame': None,
 'target_names': ['MedHouseVal'],
 'feature_names': ['MedInc',
  'HouseAge',
  'AveRooms',
  'AveBedrms',
  'Population',
  'AveOccup',
  'Latitude',
  'Longitude'],
 'DESCR': '.. _california_housing_dataset:\n

' Here we have imported the dataset from the scikit learn'

In [51]:
#pandas dataframe
df_boston = pd.DataFrame(california_housing,columns=california_housing.feature_names)

In [52]:
df_boston.head()

Empty DataFrame
Columns: [MedInc, HouseAge, AveRooms, AveBedrms, Population, AveOccup, Latitude, Longitude]
Index: []

In [53]:
# adding a column to a dataframe
df_boston["price"] = california_housing.target

In [54]:
df_boston.head()

MedInc HouseAge AveRooms AveBedrms Population AveOccup Latitude Longitude  \
0    NaN      NaN      NaN       NaN        NaN      NaN      NaN       NaN   
1    NaN      NaN      NaN       NaN        NaN      NaN      NaN       NaN   
2    NaN      NaN      NaN       NaN        NaN      NaN      NaN       NaN   
3    NaN      NaN      NaN       NaN        NaN      NaN      NaN       NaN   
4    NaN      NaN      NaN       NaN        NaN      NaN      NaN       NaN   

   price  
0  4.526  
1  3.585  
2  3.521  
3  3.413  
4  3.422

In [55]:
#removing a particular row
df_boston.drop(index=0,axis=0)

MedInc HouseAge AveRooms AveBedrms Population AveOccup Latitude  \
1        NaN      NaN      NaN       NaN        NaN      NaN      NaN   
2        NaN      NaN      NaN       NaN        NaN      NaN      NaN   
3        NaN      NaN      NaN       NaN        NaN      NaN      NaN   
4        NaN      NaN      NaN       NaN        NaN      NaN      NaN   
5        NaN      NaN      NaN       NaN        NaN      NaN      NaN   
...      ...      ...      ...       ...        ...      ...      ...   
20635    NaN      NaN      NaN       NaN        NaN      NaN      NaN   
20636    NaN      NaN      NaN       NaN        NaN      NaN      NaN   
20637    NaN      NaN      NaN       NaN        NaN      NaN      NaN   
20638    NaN      NaN      NaN       NaN        NaN      NaN      NaN   
20639    NaN      NaN      NaN       NaN        NaN      NaN      NaN   

      Longitude  price  
1           NaN  3.585  
2           NaN  3.521  
3           NaN  3.413  
4           NaN  3.422  
5           NaN  2.697  
...         ...    ...  
20635       NaN  0.781  
20636       NaN  0.771  
20637       NaN  0.923  
20638       NaN  0.847  
20639       NaN  0.894  

[20639 rows x 9 columns]

In [57]:
#removing a particular column
df_boston.drop(columns='MedInc',axis=1)

HouseAge AveRooms AveBedrms Population AveOccup Latitude Longitude  \
0          NaN      NaN       NaN        NaN      NaN      NaN       NaN   
1          NaN      NaN       NaN        NaN      NaN      NaN       NaN   
2          NaN      NaN       NaN        NaN      NaN      NaN       NaN   
3          NaN      NaN       NaN        NaN      NaN      NaN       NaN   
4          NaN      NaN       NaN        NaN      NaN      NaN       NaN   
...        ...      ...       ...        ...      ...      ...       ...   
20635      NaN      NaN       NaN        NaN      NaN      NaN       NaN   
20636      NaN      NaN       NaN        NaN      NaN      NaN       NaN   
20637      NaN      NaN       NaN        NaN      NaN      NaN       NaN   
20638      NaN      NaN       NaN        NaN      NaN      NaN       NaN   
20639      NaN      NaN       NaN        NaN      NaN      NaN       NaN   

       price  
0      4.526  
1      3.585  
2      3.521  
3      3.413  
4      3.422  
...      ...  
20635  0.781  
20636  0.771  
20637  0.923  
20638  0.847  
20639  0.894  

[20640 rows x 8 columns]

In [58]:
#locating a row using the index value
df_boston.iloc[2] # this gives second row value

MedInc          NaN
HouseAge        NaN
AveRooms        NaN
AveBedrms       NaN
Population      NaN
AveOccup        NaN
Latitude        NaN
Longitude       NaN
price         3.521
Name: 2, dtype: object

In [59]:
# locating a particular column value
print(df_boston.iloc[:,0]) # first column
print(df_boston.iloc[:,1]) # second column
print(df_boston.iloc[:,2]) # third column
print(df_boston.iloc[:,-1]) #last column   

0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
        ... 
20635    NaN
20636    NaN
20637    NaN
20638    NaN
20639    NaN
Name: MedInc, Length: 20640, dtype: object
0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
        ... 
20635    NaN
20636    NaN
20637    NaN
20638    NaN
20639    NaN
Name: HouseAge, Length: 20640, dtype: object
0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
        ... 
20635    NaN
20636    NaN
20637    NaN
20638    NaN
20639    NaN
Name: AveRooms, Length: 20640, dtype: object
0        4.526
1        3.585
2        3.521
3        3.413
4        3.422
         ...  
20635    0.781
20636    0.771
20637    0.923
20638    0.847
20639    0.894
Name: price, Length: 20640, dtype: float64


### Correlation

correlation

1. Positive correlation
2. Negative correlation

here if we take the boston dataset we can see price and crim rate a negatively correlated

if no of rooms are more price is more and it is positively correlated

In [2]:
import pandas as pd

In [3]:
#finding the number of rows and columns
df = pd.read_csv(r"C:\Users\bhara\Downloads\boston.csv")

In [4]:
df.corr()

CRIM        ZN     INDUS      CHAS       NOX        RM       AGE  \
CRIM     1.000000 -0.200469  0.406583 -0.055892  0.420972 -0.219247  0.352734   
ZN      -0.200469  1.000000 -0.533828 -0.042697 -0.516604  0.311991 -0.569537   
INDUS    0.406583 -0.533828  1.000000  0.062938  0.763651 -0.391676  0.644779   
CHAS    -0.055892 -0.042697  0.062938  1.000000  0.091203  0.091251  0.086518   
NOX      0.420972 -0.516604  0.763651  0.091203  1.000000 -0.302188  0.731470   
RM      -0.219247  0.311991 -0.391676  0.091251 -0.302188  1.000000 -0.240265   
AGE      0.352734 -0.569537  0.644779  0.086518  0.731470 -0.240265  1.000000   
DIS     -0.379670  0.664408 -0.708027 -0.099176 -0.769230  0.205246 -0.747881   
RAD      0.625505 -0.311948  0.595129 -0.007368  0.611441 -0.209847  0.456022   
TAX      0.582764 -0.314563  0.720760 -0.035587  0.668023 -0.292048  0.506456   
PTRATIO  0.289946 -0.391679  0.383248 -0.121515  0.188933 -0.355501  0.261515   
B       -0.385064  0.175520 -0.356977  0.048788 -0.380051  0.128069 -0.273534   
LSTAT    0.455621 -0.412995  0.603800 -0.053929  0.590879 -0.613808  0.602339   
MEDV    -0.388305  0.360445 -0.483725  0.175260 -0.427321  0.695360 -0.376955   

              DIS       RAD       TAX   PTRATIO         B     LSTAT      MEDV  
CRIM    -0.379670  0.625505  0.582764  0.289946 -0.385064  0.455621 -0.388305  
ZN       0.664408 -0.311948 -0.314563 -0.391679  0.175520 -0.412995  0.360445  
INDUS   -0.708027  0.595129  0.720760  0.383248 -0.356977  0.603800 -0.483725  
CHAS    -0.099176 -0.007368 -0.035587 -0.121515  0.048788 -0.053929  0.175260  
NOX     -0.769230  0.611441  0.668023  0.188933 -0.380051  0.590879 -0.427321  
RM       0.205246 -0.209847 -0.292048 -0.355501  0.128069 -0.613808  0.695360  
AGE     -0.747881  0.456022  0.506456  0.261515 -0.273534  0.602339 -0.376955  
DIS      1.000000 -0.494588 -0.534432 -0.232471  0.291512 -0.496996  0.249929  
RAD     -0.494588  1.000000  0.910228  0.464741 -0.444413  0.488676 -0.381626  
TAX     -0.534432  0.910228  1.000000  0.460853 -0.441808  0.543993 -0.468536  
PTRATIO -0.232471  0.464741  0.460853  1.000000 -0.177383  0.374044 -0.507787  
B        0.291512 -0.444413 -0.441808 -0.177383  1.000000 -0.366087  0.333461  
LSTAT   -0.496996  0.488676  0.543993  0.374044 -0.366087  1.000000 -0.737663  
MEDV     0.249929 -0.381626 -0.468536 -0.507787  0.333461 -0.737663  1.000000

Corr tell us which columns are related to each other and impact of it